### ISSUE TITLE
Proposal for more sane window splitting ex commands

### ISSUE THREAD

**Issue by Asheq** [vim/vim#13340]
I love vim's command line and its compos-ability. This feature request, in particular, is focused on making the window splitting ex commands more "sane".

Below is a table where:

- The columns are examples of actions
- The rows are various windows in which the result of those actions can be shown
  - current window
  - new horizontal window
  - new vertical window
  - new tab

(I'm using the shortest possible command names below, sometimes putting optional characters in square brackets to help readers identify naming consistencies).

 # Current
![image](https://github.com/vim/vim/assets/4657140/a3b3cd5e-ac48-4559-a35b-4d3902f615d3)
It's somewhat inconsistent -- the result of decades of features being added incrementally.

# My Proposal
![image](https://github.com/vim/vim/assets/4657140/8eeff925-3f0a-4b3e-a975-1fd5f8fae272)
Please take a look closely and compare with the current table above. Notice how much more consistent it is.

In this proposal:
- `:hw` means "horizontal window"
- `:vw` means "vertical window"
- `:tw` means "tab window"

Would love to hear your thoughts.

EDIT: Using new commands as suggested by @errael below.

**Comment by bfrg** [comment]
Just FYI: There's also `:stag` which automatically splits, so you don't have to run `:split | tag {name}`. With that, you can run `:vert stag {name}` or `:tab stag {name}`.

I think the pattern is: all commands that display a special buffer (terminal, quickfix, help) always split the current window. `:terminal` is an exception and has the `++curwin` option which will open the terminal buffer in the current window. For `:help` or `:copen` there's no equivalent option. AFAIR someone asked how to open `:help` in the current window a few months ago but I can't find the issue right now. Maybe it was on the mailing list.

All commands that edit a buffer, or jump to a location in a buffer _in the current window_, have siblings that do the same thing _in a new split_. Other commands you didn't mention are `:tselect`/`:stselect`, `:tjump`/`:stjump`, `:enew`/`:new`/`:tabnew`. And then there's also `:tabfind` and `:tabedit` which are short versions of `:tab sfind` and `:tab split`, respectively.

Yes, there is some inconsistency. Nevertheless, I highly doubt that everyone would be on board changing the current behavior and introducing new command-modifiers. So many plugins would stop working.

**Comment by errael** [comment]
I'm a lightweight when it comes to all the commands and options; but I know compatibility is critical. But consistency is always good for the newbie and lightweight user.

I notice that `:hw` and `:vw` are not in use; if you can find something to express `tab`...

**Comment by Asheq** [comment]
That's a good way to preserve backwards-compatibility. With `:hw` and `:vw`, so many commands could be deprecated, and slowly removed from use over time.

> f you can find something to express `tab`...

May be `:tw` for "tab window"

**Comment by Asheq** [comment]
> Just FYI: There's also `:stag` which automatically splits, so you don't have to run `:split | tag {name}`. With that, you can run `:vert stag {name}` or `:tab stag {name}`.

Thanks, I updated the original post.

> I think the pattern is: all commands that display a special buffer (terminal, quickfix, help) always split the current window. `:terminal` is an exception and has the `++curwin` option which will open the terminal buffer in the current window. For `:help` or `:copen` there's no equivalent option. AFAIR someone asked how to open `:help` in the current window a few months ago but I can't find the issue right now. Maybe it was on the mailing list.

I understand that the windows for these "special" buffers work in unique ways in an attempt to be helpful, but I'm not convinced that breaking consistency is worth it. And honestly, it sometimes feels more annoying than helpful.

> All commands that edit a buffer, or jump to a location in a buffer _in the current window_, have siblings that do the same thing _in a new split_. Other commands you didn't mention are `:tselect`/`:stselect`, `:tjump`/`:stjump`, `:enew`/`:new`/`:tabnew`. And then there's also `:tabfind` and `:tabedit` which are short versions of `:tab sfind` and `:tab split`, respectively.

Thanks, I added an `:enew` column. (I left out `:tselect` and `:tjump` columns because, thankfully, they seem to be very symmetrical to the existing `:tag` column)

> Yes, there is some inconsistency. Nevertheless, I highly doubt that everyone would be on board changing the current behavior and introducing new command-modifiers. So many plugins would stop working.

That's understandable. Any change would need to be backwards compatible. This could be done with new commands like @errael  suggested. Wouldn't introducing new command-modifiers be backwards compatible as well?

**Comment by bfrg** [comment]
I mentioned `:terminal ++curwin` because the command is different from the other commands and could be added to your table as well.

- `:term` opens a new terminal buffer in a (horizontally) split window
- `:vert term` in a vertical split
- `:tab term` in a new tab page, and
-  `:term ++curwin` in the current window

`++curwin` is not a command-modifier, it's an option passed to the command. `:hw`, `:vw`, `:tw` would be new command-modifiers. The current list can be found in `:h <mod>`.

**Comment by benknoble** [comment]
You’re also missing `tabedit` (which I belatedly see was pointed out already).

FWIW, especially with commands actually spelled out, the first table is at least internally consistent: split and s-prefix; vertical modifier; tab modifier. (There are lots of other modifiers, though only a few others that affect split windows.) Perhaps the only parts that doesn’t completely follow this pattern are `new` and `copen`/`cwindow`. The former I have no explanation for; the latter is special as noted because it always splits.

Really there’s probably too much inertia: to make this work, it seems to me you’d need to propose quite a large design and get approval on it. Git seems to have successfully done this with switch and restore, but I know many people (myself included, all with very different years of Git experience) who still don’t use either.

**Comment by yegappan** [comment]
Note that the [`switchbuf`](https://vimhelp.org/options.txt.html#%27switchbuf%27) option setting also impacts which window is used by some of these commands.

**Comment by Asheq** [comment]
> I mentioned `:terminal ++curwin` because the command is different from the other commands and could be added to your table as well.
> 
>     * `:term` opens a new terminal buffer in a (horizontally) split window
> 
>     * `:vert term` in a vertical split
> 
>     * `:tab term` in a new tab page, and
> 
>     * `:term ++curwin` in the current window

Thanks, I added a `:term` column.

**Comment by Asheq** [comment]
> FWIW, especially with commands actually spelled out, the first table is at least internally consistent: split and s-prefix; vertical modifier; tab modifier. (There are lots of other modifiers, though only a few others that affect split windows.) Perhaps the only parts that doesn’t completely follow this pattern are `new` and `copen`/`cwindow`. The former I have no explanation for; the latter is special as noted because it always splits.

You're right. If I remove what I'll call "shortcut helper commands" (`:vsplit`, `:tabedit`, `:vnew`, `:tabnew`, `:tabfind`), and gray out actions/columns related to "special" buffers (terminal, help, quickfix), then a somewhat consistent pattern emerges (highlighted in green). Like you said, each action has some split command (`:xyz`), vertical modifier (`:vert xyz`), and tab modifier (`:tab xyz`):

![image](https://github.com/vim/vim/assets/4657140/54263bea-538f-4069-87b3-59c37d5ca19d)

Like you said, whether the split command is the same as the base command with an "s" prefix is still inconsistent. Some are like this, e.g., `:fin`/`:sfin`, `:tag`/`:stag`, while others are not, e.g,. `:enew`/`:new`, `:edit`/`:split`.

**Comment by Asheq** [comment]
^ To continue from above, it seems we are missing some expected "shortcut helper commands":
- `:tabsp`
- `:vfind`
- `:vtag`
- `:tabtag`

Again, I'm ignoring "special" buffers (terminal, help, quickfix).

**Comment by ychin** [comment]
Just wanted to point out my old PR that proposes adding a `:curwin` modifier to Vim (#3251). Feel free to take a look. I spent a decent amount of work on it but Bram rejected the PR as not enough people (as in, no one) commented on it, but it would add universal `:curwin` (shortcut `:cur`) modifier to commands that would split a window. So for example, `:curwin terminal` would be equivalent to `:term ++curwin`, but `:cur split` is also equivalent to `:edit`, and it also works with `:help` and `:copen`.

Furthermore, it allows you to add a window number to put the window where you want, so `:+1curwin copen` would open the quickfix window at the window directly after the current window, `:2curwin help` will put the help window at the 2nd window in the current tab.

I think this results in the most flexible and intuitive way to solve this. FWIW I don't think changing *any* defaults in this is a good idea due to the vast amount of scripts and user expectations of what these commands do, so e.g. the original proposal of making `:help` and `:terminal` open a window in place doesn't seem like a great idea to me. I maintained a separate branch with a working `:curwin` modifier for a few years until eventually I stopped.

Another motivation for having `:curwin` be able to specify a window number (instead of just modifying the current window) is to make it easy to have session restore (from one made with `:mksession`) capable of restoring quickfix windows (it currently doesn't have the ability to do that).

> AFAIR someone asked how to open :help in the current window a few months ago but I can't find the issue right now. Maybe it was on the mailing list

I don't think this is possible. This is why I made that PR.

---

Separate from that, I'm trying to understand what this is proposing. Is `:tw` just a shortcut for `:tab`, except if you use it raw it will be the equivalent to `:tab split`? I'm not really sure if I like that because a like modifiers being modifiers instead of "can be used as both a command and a modifier". Maybe I'm missing something here. And I feel that `:tab` is easier to remember than `:tw`.
